File: 18-heaps/5-top-k-problems/main.cpp
// Top K Elements - C++ Example

#include <iostream>
#include <vector>
#include <queue>

std::vector<int> findTopKElements(const std::vector<int>& arr, int k) {
    if (k <= 0) return {};
    std::priority_queue<int, std::vector<int>, std::greater<int>> minHeap;
    for (int num : arr) {
        if (minHeap.size() < k) {
            minHeap.push(num);
        } else if (num > minHeap.top()) {
            minHeap.pop();
            minHeap.push(num);
        }
    }
    std::vector<int> result;
    while (!minHeap.empty()) {
        result.push_back(minHeap.top());
        minHeap.pop();
    }
    return result;
}

int main() {
    std::vector<int> arr = {3,1,4,1,5,9,2,6,5};
    int k = 3;
    std::vector<int> topK = findTopKElements(arr, k);
    std::cout << "Top " << k << " elements are: ";
    for (int val : topK) std::cout << val << " ";
    std::cout << std::endl;
    return 0;
}

File: 18-heaps/5-top-k-problems/main.java
// Top K Elements - Java Example

import java.util.PriorityQueue;

class TopKElements {

    public static int[] findTopKElements(int[] arr, int k) {
        if (k <= 0) return new int[0];
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for (int num : arr) {
            if (minHeap.size() < k) {
                minHeap.add(num);
            } else if (num > minHeap.peek()) {
                minHeap.poll();
                minHeap.add(num);
            }
        }
        int[] result = new int[minHeap.size()];
        int i = 0;
        for (int val : minHeap)
            result[i++] = val;
        return result;
    }

    public static void main(String[] args) {
        int[] arr = {3, 1, 4, 1, 5, 9, 2, 6, 5};
        int k = 3;
        int[] topK = findTopKElements(arr, k);
        System.out.print("Top " + k + " elements are: ");
        for (int val : topK)
            System.out.print(val + " ");
    }
}
